<a href="https://colab.research.google.com/github/Nareekan789/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [3]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/bsc_dpdm23_data'

Mounted at /content/drive


In [4]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/dsc_dpdm23_data/reduced_marketbasket.csv'),encoding='latin-1')

In [6]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

###Quiz:ข้อมูลมาจากกี่ประเทศ

In [11]:
len(set['data'])

TypeError: object of type 'types.GenericAlias' has no len()

# แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย groupby.apply(list)

In [ ]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

In [13]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [14]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=844b72815081535da9ecc4861dc8177fdf6948c7d9383d82801390a676f9b7a6
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [15]:
from apyori import apriori

In [16]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [17]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [18]:
len(frequentItemsets)

4

In [19]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [20]:
frequentItemsets[0][0]

frozenset({'22423'})

In [21]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [22]:
frequentItemsets[0][0].__len__()

1

In [23]:
list(apriori(transactions,min_support = 0.03))

[RelationRecord(items=frozenset({'20712'}), support=0.033667953667953665, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.033667953667953665, lift=1.0)]),
 RelationRecord(items=frozenset({'20719'}), support=0.03254826254826255, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719'}), confidence=0.03254826254826255, lift=1.0)]),
 RelationRecord(items=frozenset({'20724'}), support=0.04054054054054054, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20724'}), confidence=0.04054054054054054, lift=1.0)]),
 RelationRecord(items=frozenset({'20725'}), support=0.062084942084942085, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.062084942084942085, lift=1.0)]),
 RelationRecord(items=frozenset({'20726'}), support=0.04003861003861004, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozense

In [24]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [25]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

## quiz 22386 และ 85099B คือสินค้าอะไร

In [26]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [27]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'537823',
 '555854',
 '578936',
 '556083',
 '555150',
 '542583',
 '541008',
 '560773',
 '565150',
 '579927',
 '545526',
 '557605',
 'C541822',
 '579196',
 '544456',
 '551340',
 'C539071',
 '554798',
 '579518',
 '553891',
 '571745',
 '551953',
 'C548355',
 '575875',
 '569650',
 '554039',
 '539217',
 '558315',
 '540777',
 '553461',
 '541866',
 '550468',
 '580610',
 '580983',
 '544596',
 '540915',
 '575846',
 '564754',
 '548714',
 '580053',
 '543022',
 '548514',
 '569834',
 '564432',
 '553146',
 '540539',
 '569246',
 '568346',
 '554337',
 '547096',
 '565155',
 '545320',
 '544824',
 '542999',
 '541932',
 '542389',
 '545601',
 '544301',
 '539752',
 '542779',
 '547360',
 '560790',
 '580730',
 'C566758',
 '577594',
 '549360',
 '552323',
 '574731',
 '542719',
 '557788',
 '567922',
 '571713',
 '580727',
 'C562375',
 '544075',
 '576616',
 '561089',
 '568651',
 'C548485',
 '550307',
 '552678',
 '561645',
 '576181',
 '537042',
 '570514',
 '554382',
 '578943',
 '554694',
 '537022',
 '558774',
 '54

In [28]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

### HW 1  หา Frequent  Item Set และ Association Rules
- ของช่วงเวลา christmas & New Year (1 Dec - 31 Dec)
- ของประเทศในทวีปยุโรป
- ของช่วงเวลา ตรุษจีน ในทวีป Asia(15 Jan - 15 Feb)
- ของช่วงเวลา christmas & New Year ในทวีปยุโรป (1 Dec - 31 Dec)
- ของวันวาเลนไทน์
- โดยจัดกลุ่มตามประเทศ

###เตรียมข้อมูล

ข้อมูลมาจากกี่ประเทศ

In [29]:
data.groupby(['Country']).sum()

<ipython-input-29-ca3bc1201a7e>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  data.groupby(['Country']).sum()


Quantity    UnitPrice    CustomerID
Country                                                  
Australia                83653     4054.750  1.569300e+07
Austria                   4827     1701.520  5.021102e+06
Bahrain                    260       86.570  2.100270e+05
Belgium                  23152     7540.130  2.571829e+07
Brazil                     356      142.600  4.086080e+05
Canada                    2763      910.580  2.615483e+06
Channel Islands           9479     3738.550  1.128522e+07
Cyprus                    6317     3920.070  7.715880e+06
Czech Republic             592       88.150  3.834300e+05
Denmark                   8188     1266.950  4.876734e+06
EIRE                    142637    48447.190  1.103917e+08
European Community         497      294.050  9.215880e+05
Finland                  10666     3786.850  8.699324e+06
France                  110480    43031.990  1.076489e+08
Germany                 117448    37666.000  1.200751e+08
Greece                    1556      713.290  2.008584e+06
Hong Kong                 4769    12241.500  0.000000e+00
Iceland                   2458      481.210  2.247154e+06
Israel                    4353     1079.040  3.164467e+06
Italy                     7999     3879.390  1.015666e+07
Japan                    25218      814.860  4.567292e+06
Lebanon                    386      242.440  5.743800e+05
Lithuania                  652       99.440  5.366200e+05
Malta                      944      666.010  2.158496e+06
Netherlands             200128     6492.550  3.419054e+07
Norway                   19247     6529.060  1.350765e+07
Poland                    3653     1422.270  4.341972e+06
Portugal                 16180    13037.540  1.886480e+07
RSA                        352      248.100  7.218680e+05
Saudi Arabia                75       24.110  1.256500e+05
Singapore                 5234    25108.890  2.918376e+06
Spain                    26824    12633.450  3.268929e+07
Sweden                   35637     1806.830  6.790083e+06
Switzerland              30325     6813.690  2.377592e+07
USA                       1034      644.980  3.672086e+06
United Arab Emirates       982      229.890  1.018952e+06
United Kingdom         4263829  2245715.474  5.626433e+09
Unspecified               3300     1204.010  3.348046e+06

แปลงdata ให้อยู่ในรูปtransaction

In [30]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [31]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

Association Rules

In [32]:
!pip install apyori


In [33]:
from apyori import apriori

In [35]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [37]:
len(frequentItemsets)

4

In [38]:
frequentItemsets[0][0]

frozenset({'22423'})

In [39]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [40]:
frequentItemsets[0][0].__len__()

1

In [41]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


In [42]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

In [43]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'537823',
 '555854',
 '578936',
 '556083',
 '555150',
 '542583',
 '541008',
 '560773',
 '565150',
 '579927',
 '545526',
 '557605',
 'C541822',
 '579196',
 '544456',
 '551340',
 'C539071',
 '554798',
 '579518',
 '553891',
 '571745',
 '551953',
 'C548355',
 '575875',
 '569650',
 '554039',
 '539217',
 '558315',
 '540777',
 '553461',
 '541866',
 '550468',
 '580610',
 '580983',
 '544596',
 '540915',
 '575846',
 '564754',
 '548714',
 '580053',
 '543022',
 '548514',
 '569834',
 '564432',
 '553146',
 '540539',
 '569246',
 '568346',
 '554337',
 '547096',
 '565155',
 '545320',
 '544824',
 '542999',
 '541932',
 '542389',
 '545601',
 '544301',
 '539752',
 '542779',
 '547360',
 '560790',
 '580730',
 'C566758',
 '577594',
 '549360',
 '552323',
 '574731',
 '542719',
 '557788',
 '567922',
 '571713',
 '580727',
 'C562375',
 '544075',
 '576616',
 '561089',
 '568651',
 'C548485',
 '550307',
 '552678',
 '561645',
 '576181',
 '537042',
 '570514',
 '554382',
 '578943',
 '554694',
 '537022',
 '558774',
 '54

###ข้อ1 ช่วงเวลา Christmas and New year (1 Dec - 31 Dec)

In [44]:
import datetime

In [45]:
data['InvoiceDate']

0         12/01/2010 08:26
1         12/01/2010 08:26
2         12/01/2010 08:26
3         12/01/2010 08:26
4         12/01/2010 08:26
                ...       
541904    12/09/2011 12:50
541905    12/09/2011 12:50
541906    12/09/2011 12:50
541907    12/09/2011 12:50
541908    12/09/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

In [46]:
data['InvoiceDate']=pd.to_datetime(data.InvoiceDate, format='%m/%d/%Y %H:%M' )

In [47]:
data['Date'] = data['InvoiceDate'].dt.date
data['Date']=pd.to_datetime(data.Date, format='%Y/%m/%d')
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09  

[541909 rows x 9 columns]

In [48]:
#Select 1-31 Dec
December_1 = data[(data['Date'].dt.month == 12) & (data['Date'].dt.day >= 1) & (data['Date'].dt.day <= 31)]
#ปริ้นเพื่อตรวจสอบ
print(December_1)


       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Dat

In [49]:
transactions1 = December_1.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions1)

     InvoiceNo                                        Description
0       536365  [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       536366  [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       536367  [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       536368  [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4       536369                         [BATH BUILDING BLOCK WORD]
...        ...                                                ...
3035   C581484                      [PAPER CRAFT , LITTLE BIRDIE]
3036   C581490  [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3037   C581499                                           [Manual]
3038   C581568                       [VICTORIAN SEWING BOX LARGE]
3039   C581569  [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...

[3040 rows x 2 columns]


In [50]:
transactions1_apply = transactions1['Description'].apply(lambda x: [str(item) for item in x])
transactions1_apply

0       [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                              [BATH BUILDING BLOCK WORD]
                              ...                        
3035                        [PAPER CRAFT , LITTLE BIRDIE]
3036    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3037                                             [Manual]
3038                         [VICTORIAN SEWING BOX LARGE]
3039    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 3040, dtype: object

In [51]:

frequentItemset = list(apriori(transactions1_apply, min_support=0.036))

for record in frequentItemset:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()



yes
Items: ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED 
Support: 0.0368 (3.68%)
Rule 1: (None -> ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED  : Confidence=0.0368, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7089, Lift=14.4627)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7517, Lift=14.4627)

yes
Items: HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE
Support: 0.0365 (3.65%)
Rule 1: (None -> HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE : Confidence=0.0365, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7484)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7484)

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0477 (4.77%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS : 

###ข้อ2 ของประเทศในทวีปยุโรป

In [54]:
europe = ['Austria' ,'Belgium' ,'Channel Islands' ,'Cyprus' ,'Czech Republic' ,'Denmark' , 'EIRE' ,'European Community' , 'Finland' ,'France' ,'Germany' , 'Greece' , 'Iceland' , 'Italy' , 'Lithuania' , 'Malta' ,'Netherlands' , 'Norway' ,'Poland' ,'Portugal' , 'Spain' ,'Sweden' ,'Switzerland' ,'United Kingdom']

In [55]:
EU_data = data[data['Country']==""]
for i in range(len(europe)):
  x = data[data['Country']==europe[i]]
  EU_data=pd.concat([x,EU_data])

In [56]:
print(EU_data)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country       Dat

In [58]:
transactions2 = EU_data.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions2)

      InvoiceNo                                        Description
0        536365  [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1        536366  [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2        536367  [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3        536368  [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4        536369                         [BATH BUILDING BLOCK WORD]
...         ...                                                ...
25726   C581484                      [PAPER CRAFT , LITTLE BIRDIE]
25727   C581490  [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
25728   C581499                                           [Manual]
25729   C581568                       [VICTORIAN SEWING BOX LARGE]
25730   C581569  [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...

[25731 rows x 2 columns]


In [59]:
transactions2_apply = transactions2['Description'].apply(lambda x: [str(item) for item in x])
transactions2_apply

0        [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1        [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2        [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3        [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                               [BATH BUILDING BLOCK WORD]
                               ...                        
25726                        [PAPER CRAFT , LITTLE BIRDIE]
25727    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
25728                                             [Manual]
25729                         [VICTORIAN SEWING BOX LARGE]
25730    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 25731, dtype: object

In [60]:
frequentItemset2 = list(apriori(transactions2_apply, min_support=0.025))

for record in frequentItemset2:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: ROSES REGENCY TEACUP AND SAUCER , GREEN REGENCY TEACUP AND SAUCER
Support: 0.0301 (3.01%)
Rule 1: (None -> ROSES REGENCY TEACUP AND SAUCER , GREEN REGENCY TEACUP AND SAUCER : Confidence=0.0301, Lift=1.0000)
Rule 2: (GREEN REGENCY TEACUP AND SAUCER -> ROSES REGENCY TEACUP AND SAUCER  : Confidence=0.7402, Lift=17.2054)
Rule 3: (ROSES REGENCY TEACUP AND SAUCER  -> GREEN REGENCY TEACUP AND SAUCER : Confidence=0.7001, Lift=17.2054)

yes
Items: JUMBO BAG PINK POLKADOT, JUMBO BAG RED RETROSPOT
Support: 0.0323 (3.23%)
Rule 1: (None -> JUMBO BAG PINK POLKADOT, JUMBO BAG RED RETROSPOT : Confidence=0.0323, Lift=1.0000)
Rule 2: (JUMBO BAG PINK POLKADOT -> JUMBO BAG RED RETROSPOT : Confidence=0.6775, Lift=8.2039)
Rule 3: (JUMBO BAG RED RETROSPOT -> JUMBO BAG PINK POLKADOT : Confidence=0.3915, Lift=8.2039)

yes
Items: JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAGE RED PAISLEY
Support: 0.0265 (2.65%)
Rule 1: (None -> JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAGE RED PAISLEY : Confidence=0

###ข้อ 3 ของช่วงเวลาตรุษจันในทวีปเอเชีย 15Jan-15Feb

In [62]:
Asia = ['Bahrain' , 'Hong Kong' , 'Israel' , 'Japan' , 'Lebanon' , 'Saudi Arabia' , 'Singapore', 'United Arab Emirates']

In [63]:
Asia_data = data[data['Country']==""]
for i in range(len(Asia)):
  x = data[data['Country']==Asia[i]]
  Asia_data=pd.concat([x,Asia_data])
  Asia_data

In [64]:
#Chinese Newyear
China1 = Asia_data[(Asia_data['Date'].dt.month == 1) & (Asia_data['Date'].dt.day >= 15) & (data['Date'].dt.day <= 31)  |
                   ((Asia_data['Date'].dt.month == 2) & (Asia_data['Date'].dt.day <= 15))]
#ปริ้นเพื่อตรวจสอบ
print(China1)

      InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   
...         ...       ...                               ...       ...   
69675    541975     20726                LUNCH BAG WOODLAND        20   
69676    541975     22452      MEASURING TAPE BABUSHKA PINK        24   
69677    541975     22454       MEASURING TAPE BABUSHKA RED        24   
69678    541975     20749        ASSORTED COLOUR MINI CASES         6   
69679    541975     20682  RED RETROSPOT CHILDRENS UMBRELLA         6   

              InvoiceDate  UnitPrice  CustomerID               Country  \
89570 2011-02-14 12:46:00       4.95     17829.0 

<ipython-input-64-1ddecad5b434>:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  China1 = Asia_data[(Asia_data['Date'].dt.month == 1) & (Asia_data['Date'].dt.day >= 15) & (data['Date'].dt.day <= 31)  |


In [65]:
transactions3 = China1.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions3)

  InvoiceNo                                        Description
0    541975  [4 TRADITIONAL SPINNING TOPS, CHARLIE & LOLA W...
1    542102  [GIN & TONIC DIET GREETING CARD , GREEN REGENC...
2    542276  [LAUNDRY 15C METAL SIGN, BEWARE OF THE CAT MET...
3    543179  [MONEY BOX FIRST ADE DESIGN, MONEY BOX KINGS C...
4    543518  [SET OF 3 NOTEBOOKS IN PARCEL, LUNCH BAG DOLLY...
5    543911  [RETROSPOT HEART HOT WATER BOTTLE, DOORMAT UNI...


In [66]:
transactions3_apply = transactions3['Description'].apply(lambda x: [str(item) for item in x])
transactions3_apply

0    [4 TRADITIONAL SPINNING TOPS, CHARLIE & LOLA W...
1    [GIN & TONIC DIET GREETING CARD , GREEN REGENC...
2    [LAUNDRY 15C METAL SIGN, BEWARE OF THE CAT MET...
3    [MONEY BOX FIRST ADE DESIGN, MONEY BOX KINGS C...
4    [SET OF 3 NOTEBOOKS IN PARCEL, LUNCH BAG DOLLY...
5    [RETROSPOT HEART HOT WATER BOTTLE, DOORMAT UNI...
Name: Description, dtype: object

In [67]:
frequentItemset3 = list(apriori(transactions3_apply, min_support=0.33))

for record in frequentItemset3:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: LUNCH BAG RED RETROSPOT, BASKET OF TOADSTOOLS
Support: 0.3333 (33.33%)
Rule 1: (None -> LUNCH BAG RED RETROSPOT, BASKET OF TOADSTOOLS : Confidence=0.3333, Lift=1.0000)
Rule 2: (BASKET OF TOADSTOOLS -> LUNCH BAG RED RETROSPOT : Confidence=1.0000, Lift=3.0000)
Rule 3: (LUNCH BAG RED RETROSPOT -> BASKET OF TOADSTOOLS : Confidence=1.0000, Lift=3.0000)

yes
Items: SET 3 RETROSPOT TEA,COFFEE,SUGAR, BASKET OF TOADSTOOLS
Support: 0.3333 (33.33%)
Rule 1: (None -> SET 3 RETROSPOT TEA,COFFEE,SUGAR, BASKET OF TOADSTOOLS : Confidence=0.3333, Lift=1.0000)
Rule 2: (BASKET OF TOADSTOOLS -> SET 3 RETROSPOT TEA,COFFEE,SUGAR : Confidence=1.0000, Lift=3.0000)
Rule 3: (SET 3 RETROSPOT TEA,COFFEE,SUGAR -> BASKET OF TOADSTOOLS : Confidence=1.0000, Lift=3.0000)

yes
Items: JAM MAKING SET WITH JARS, BATH BUILDING BLOCK WORD
Support: 0.3333 (33.33%)
Rule 1: (None -> JAM MAKING SET WITH JARS, BATH BUILDING BLOCK WORD : Confidence=0.3333, Lift=1.0000)
Rule 2: (BATH BUILDING BLOCK WORD -> JAM MAKING SET

###ข้อ 4 ช่วงเวลาChristmas & New year ในทวีปยุโรป

In [69]:
#Select 1-31 Dec
December2 = EU_data[(EU_data['Date'].dt.month == 12) & (EU_data['Date'].dt.day >= 1) & (EU_data['Date'].dt.day <= 31)]
#ปริ้นเพื่อตรวจสอบ
print(December2)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country       Dat

In [70]:
#EU Christmas
Eu_christ = EU_data[(EU_data['Date'].dt.month == 12) & (EU_data['Date'].dt.day >= 1) & (EU_data['Date'].dt.day <= 31)]
Eu_christ

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
536964 2011-12-08 10:26:00       5.95     12358.0         Austria 2011-12-08  
536965 2011-12-08 10:26:00       5.95     12358.0         Austria 2011-12-08  
536966 2011-12-08 10:26:00       0.39     12358.0         Austria 2011-12-08  
536967 2011-12-08 10:26:00       0.39     12358.0         Austria 2011-12-08  
536968 2011-12-08 10:26:00      40.00     12358.0         Austria 2011-12-08  

[67852 rows x 9 columns]

In [71]:
transactions4 = Eu_christ.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions4)

     InvoiceNo                                        Description
0       536365  [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       536366  [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       536367  [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       536368  [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4       536369                         [BATH BUILDING BLOCK WORD]
...        ...                                                ...
3022   C581484                      [PAPER CRAFT , LITTLE BIRDIE]
3023   C581490  [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3024   C581499                                           [Manual]
3025   C581568                       [VICTORIAN SEWING BOX LARGE]
3026   C581569  [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...

[3027 rows x 2 columns]


In [72]:
transactions4_apply = transactions4['Description'].apply(lambda x: [str(item) for item in x])
transactions4_apply

0       [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                              [BATH BUILDING BLOCK WORD]
                              ...                        
3022                        [PAPER CRAFT , LITTLE BIRDIE]
3023    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3024                                             [Manual]
3025                         [VICTORIAN SEWING BOX LARGE]
3026    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 3027, dtype: object

In [73]:
#items ที่มักซื้อด้วยกัน
frequentItemset4 = list(apriori(transactions4_apply, min_support=0.015))
for record in frequentItemset4:
    if len(record.items) > 2:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")
        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift
            print(f"Rule {i}: ({items_base} -> {items_add}: Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , 60 CAKE CASES VINTAGE CHRISTMAS, PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0182 (1.82%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , 60 CAKE CASES VINTAGE CHRISTMAS, PAPER CHAIN KIT VINTAGE CHRISTMAS: Confidence=0.0182, Lift=1.0000)
Rule 2: (60 CAKE CASES VINTAGE CHRISTMAS -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS: Confidence=0.4264, Lift=8.9006)
Rule 3: (PAPER CHAIN KIT 50'S CHRISTMAS  -> 60 CAKE CASES VINTAGE CHRISTMAS, PAPER CHAIN KIT VINTAGE CHRISTMAS: Confidence=0.1957, Lift=8.5866)
Rule 4: (PAPER CHAIN KIT VINTAGE CHRISTMAS -> PAPER CHAIN KIT 50'S CHRISTMAS , 60 CAKE CASES VINTAGE CHRISTMAS: Confidence=0.2523, Lift=10.6068)
Rule 5: (PAPER CHAIN KIT 50'S CHRISTMAS , 60 CAKE CASES VINTAGE CHRISTMAS -> PAPER CHAIN KIT VINTAGE CHRISTMAS: Confidence=0.7639, Lift=10.6068)
Rule 6: (60 CAKE CASES VINTAGE CHRISTMAS, PAPER CHAIN KIT VINTAGE CHRISTMAS -> PAPER CHAIN KIT 50'S CHRISTMAS : Confidence=0.7971, Lift=8

In [74]:
frequentItemset4 = list(apriori(transactions4_apply, min_support=0.035))

for record in frequentItemset4:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED 
Support: 0.0367 (3.67%)
Rule 1: (None -> ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED  : Confidence=0.0367, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7070, Lift=14.4602)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7500, Lift=14.4602)

yes
Items: HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE
Support: 0.0367 (3.67%)
Rule 1: (None -> HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE : Confidence=0.0367, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7067)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7067)

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0479 (4.79%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS : 

###ข้อ 5 ของวันวาเลนไทน์

In [76]:
#Select 1-31 Dec
Feb14 = data[(data['Date'].dt.month == 2) & (data['Date'].dt.day == 14)]
#ปริ้นเพื่อตรวจสอบ
print(Feb14)

      InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

              InvoiceDate  UnitPrice  CustomerID         Country       Date  
88847 2011-02-14 09:2

In [77]:
transactions5 = Feb14.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transactions5)

   InvoiceNo                                        Description
0     543821  [IVORY GIANT GARDEN THERMOMETER, YELLOW GIANT ...
1     543822  [RETROSPOT TEA SET CERAMIC 11 PC , 12 PENCIL S...
2     543823  [BAKING SET 9 PIECE RETROSPOT , GLASS  BEURRE ...
3     543824  [POPPY'S PLAYHOUSE BEDROOM , POPPY'S PLAYHOUSE...
4     543825  [IVORY GIANT GARDEN THERMOMETER, SET 3 WICKER ...
5     543826  [PURPLE FRANGIPANI HAIRCLIP, FINE SILVER NECKL...
6     543827                    [incorrectly made-thrown away.]
7     543828  [SET OF 6 HERB TINS SKETCHBOOK, SET OF 6 SPICE...
8     543829  [RED RETROSPOT CUP, PINK  POLKADOT CUP, BLUE P...
9     543831  [TOMATO CHARLIE+LOLA COASTER SET, JUMBO BAG TO...
10    543832  [ASSORTED COLOUR BIRD ORNAMENT, 6 RIBBONS RUST...
11    543833  [BROCANTE COAT RACK, SET 3 PAPER VINTAGE CHICK...
12    543836  [SET OF 36 TEATIME PAPER DOILIES, 60 TEATIME F...
13    543839  [FRYING PAN UNION FLAG, RED TOADSTOOL LED NIGH...
14    543842               [GROW A FLYTR

In [78]:
transactions5_apply = transactions5['Description'].apply(lambda x: [str(item) for item in x])
transactions5_apply

0     [IVORY GIANT GARDEN THERMOMETER, YELLOW GIANT ...
1     [RETROSPOT TEA SET CERAMIC 11 PC , 12 PENCIL S...
2     [BAKING SET 9 PIECE RETROSPOT , GLASS  BEURRE ...
3     [POPPY'S PLAYHOUSE BEDROOM , POPPY'S PLAYHOUSE...
4     [IVORY GIANT GARDEN THERMOMETER, SET 3 WICKER ...
5     [PURPLE FRANGIPANI HAIRCLIP, FINE SILVER NECKL...
6                       [incorrectly made-thrown away.]
7     [SET OF 6 HERB TINS SKETCHBOOK, SET OF 6 SPICE...
8     [RED RETROSPOT CUP, PINK  POLKADOT CUP, BLUE P...
9     [TOMATO CHARLIE+LOLA COASTER SET, JUMBO BAG TO...
10    [ASSORTED COLOUR BIRD ORNAMENT, 6 RIBBONS RUST...
11    [BROCANTE COAT RACK, SET 3 PAPER VINTAGE CHICK...
12    [SET OF 36 TEATIME PAPER DOILIES, 60 TEATIME F...
13    [FRYING PAN UNION FLAG, RED TOADSTOOL LED NIGH...
14                 [GROW A FLYTRAP OR SUNFLOWER IN TIN]
15    [CHARLOTTE BAG SUKI DESIGN, JUMBO STORAGE BAG ...
16    [CARRIAGE, CLASSIC METAL BIRDCAGE PLANT HOLDER...
17    [CHARLOTTE BAG SUKI DESIGN, WOODLAND CHARL

In [79]:
frequentItemset5 = list(apriori(transactions5_apply, min_support=0.09))

for record in frequentItemset5:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: HOMEMADE JAM SCENTED CANDLES, SET OF 3 CAKE TINS PANTRY DESIGN 
Support: 0.0909 (9.09%)
Rule 1: (None -> HOMEMADE JAM SCENTED CANDLES, SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=0.0909, Lift=1.0000)
Rule 2: (HOMEMADE JAM SCENTED CANDLES -> SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=1.0000, Lift=6.8750)
Rule 3: (SET OF 3 CAKE TINS PANTRY DESIGN  -> HOMEMADE JAM SCENTED CANDLES : Confidence=0.6250, Lift=6.8750)

yes
Items: SET OF 6 SPICE TINS PANTRY DESIGN, SET OF 3 CAKE TINS PANTRY DESIGN 
Support: 0.0909 (9.09%)
Rule 1: (None -> SET OF 6 SPICE TINS PANTRY DESIGN, SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=0.0909, Lift=1.0000)
Rule 2: (SET OF 3 CAKE TINS PANTRY DESIGN  -> SET OF 6 SPICE TINS PANTRY DESIGN : Confidence=0.6250, Lift=5.7292)
Rule 3: (SET OF 6 SPICE TINS PANTRY DESIGN -> SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=0.8333, Lift=5.7292)



###ข้อ6 จัดกลุ่มตามประเทศและเดือน

In [81]:
data['InvoiceDate']=pd.to_datetime(data.InvoiceDate, format='%m/%d/%Y %H:%M' )
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09  

[541909 rows x 9 columns]

In [82]:
data['Month'] = data['InvoiceDate'].dt.month
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01   
...                    ...        ...         ...             ...        ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09   
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09   
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09   
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09   
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09   

        Month  
0          12  
1          12  
2          12  
3          12  
4          12  
...       ...  
541904     12  
541905     12  
541906     12  
541907     12  
541908     12  

[541909 rows x 10 columns]

In [83]:
transWorld = data.groupby(['InvoiceNo','Country', 'Month'])['StockCode'].apply(list)
transWorld

InvoiceNo  Country         Month
536365     United Kingdom  12       [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366     United Kingdom  12                                          [22633, 22632]
536367     United Kingdom  12       [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368     United Kingdom  12                            [22960, 22913, 22912, 22914]
536369     United Kingdom  12                                                 [21756]
                                                          ...                        
C581484    United Kingdom  12                                                 [23843]
C581490    United Kingdom  12                                          [22178, 23144]
C581499    United Kingdom  12                                                     [M]
C581568    United Kingdom  12                                                 [21258]
C581569    United Kingdom  12                                          [84978, 20979]
Name: StockCode, Leng

In [84]:
min_sup = 0.025
frequentItemsets6 = list(apriori(transWorld,min_support = min_sup))
for i in frequentItemsets6:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'20725', '20727'}), support=0.025019305019305018, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725', '20727'}), confidence=0.025019305019305018, lift=1.0), OrderedStatistic(items_base=frozenset({'20725'}), items_add=frozenset({'20727'}), confidence=0.40298507462686567, lift=8.059701492537313), OrderedStatistic(items_base=frozenset({'20727'}), items_add=frozenset({'20725'}), confidence=0.5003861003861003, lift=8.059701492537313)])
yes
RelationRecord(items=frozenset({'20725', '22383'}), support=0.0255984555984556, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725', '22383'}), confidence=0.0255984555984556, lift=1.0), OrderedStatistic(items_base=frozenset({'20725'}), items_add=frozenset({'22383'}), confidence=0.41231343283582095, lift=8.176813101414826), OrderedStatistic(items_base=frozenset({'22383'}), items_add=frozenset({'20725'}), confidence=0.5076569678407351, lift=8.176

In [85]:
set(data[(data['StockCode']=='85099B') | (data['StockCode']=='22386')]['Description']) #มากสุด , support=0.03216216216216216. confidence=0.03216216216216216, lift=1.0)

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

In [86]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['Description']) #อันดับ 2 support=0.03027027027027027 confidence=0.03027027027027027, lift=1.0

{'GREEN REGENCY TEACUP AND SAUCER', 'ROSES REGENCY TEACUP AND SAUCER '}

In [87]:
set(data[(data['StockCode']=='21931') | (data['StockCode']=='85099B')]['Description']) # อันดับ 3 support=0.0283011583011583 confidence=0.0283011583011583, lift=1.0

{'JUMBO BAG RED RETROSPOT', 'JUMBO STORAGE BAG SUKI'}

In [88]:
transWorld101 = data.groupby(['InvoiceNo','Country', 'Month'])['StockCode'].apply(list).reset_index()
transWorld101

InvoiceNo         Country  Month  \
0        536365  United Kingdom     12   
1        536366  United Kingdom     12   
2        536367  United Kingdom     12   
3        536368  United Kingdom     12   
4        536369  United Kingdom     12   
...         ...             ...    ...   
25895   C581484  United Kingdom     12   
25896   C581490  United Kingdom     12   
25897   C581499  United Kingdom     12   
25898   C581568  United Kingdom     12   
25899   C581569  United Kingdom     12   

                                               StockCode  
0      [85123A, 71053, 84406B, 84029G, 84029E, 22752,...  
1                                         [22633, 22632]  
2      [84879, 22745, 22748, 22749, 22310, 84969, 226...  
3                           [22960, 22913, 22912, 22914]  
4                                                [21756]  
...                                                  ...  
25895                                            [23843]  
25896                                     [22178, 23144]  
25897                                                [M]  
25898                                            [21258]  
25899                                     [84978, 20979]  

[25900 rows x 4 columns]

In [89]:
transWorld101_apply = transWorld101['StockCode'].apply(lambda x: [str(item) for item in x])
transWorld101_apply

0        [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
1                                           [22633, 22632]
2        [84879, 22745, 22748, 22749, 22310, 84969, 226...
3                             [22960, 22913, 22912, 22914]
4                                                  [21756]
                               ...                        
25895                                              [23843]
25896                                       [22178, 23144]
25897                                                  [M]
25898                                              [21258]
25899                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [90]:
frequentItemset7 = list(apriori(transWorld101_apply, min_support=0.028))

for record in frequentItemset7:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: 85099B, 21931
Support: 0.0283 (2.83%)
Rule 1: (None -> 85099B, 21931 : Confidence=0.0283, Lift=1.0000)
Rule 2: (21931 -> 85099B : Confidence=0.6103, Lift=7.4039)
Rule 3: (85099B -> 21931 : Confidence=0.3433, Lift=7.4039)

yes
Items: 22386, 85099B
Support: 0.0322 (3.22%)
Rule 1: (None -> 22386, 85099B : Confidence=0.0322, Lift=1.0000)
Rule 2: (22386 -> 85099B : Confidence=0.6767, Lift=8.2090)
Rule 3: (85099B -> 22386 : Confidence=0.3902, Lift=8.2090)

yes
Items: 22697, 22699
Support: 0.0303 (3.03%)
Rule 1: (None -> 22697, 22699 : Confidence=0.0303, Lift=1.0000)
Rule 2: (22697 -> 22699 : Confidence=0.7417, Lift=17.1523)
Rule 3: (22699 -> 22697 : Confidence=0.7000, Lift=17.1523)

